src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerManager.cs

```diff
@@ -377,11 +378,13 @@ internal bool IsSupportedDiagnostic(DiagnosticAnalyzer analyzer, Diagnostic diag
                 SeverityFilter severityFilter,
                 Compilation? compilation,
                 AnalyzerOptions? analyzerOptions,
+                AnalysisScope? analysisScope,
                 CancellationToken cancellationToken)
             {
                 if (compilation != null && compilation.Options.SyntaxTreeOptionsProvider is { } treeOptions)
                 {
-                    foreach (var tree in compilation.SyntaxTrees)
+                    var trees = analysisScope?.SyntaxTrees ?? compilation.SyntaxTrees;
+                    foreach (var tree in trees)
```

Comment:
just to check, but this is primary perf win right?  Because the analysis scope can provide a smaller set of trees, we don't have to check so many trees to determine if the the diagnostic analyzer is suppressed or not?

And, i'm assumign on the IDE side, we create an AnalysisScope limited to the current file?

Reply:
> just to check, but this is primary perf win right? Because the analysis scope can provide a smaller set of trees, we don't have to check so many trees to determine if the the diagnostic analyzer is suppressed or not?

Yes, correct. The analysis scope will just have a single tree for IDE single-file analysis case.

> And, i'm assumign on the IDE side, we create an AnalysisScope limited to the current file?

Yes, we invoke the below public API for IDE single-file semantic analysis, which internally creates and passes in an AnalysisScope with just this one tree: https://github.com/dotnet/roslyn/blob/3fb0657283726a1436d33370c41372fac51c56f5/src/Compilers/Core/Portable/DiagnosticAnalyzer/CompilationWithAnalyzers.cs#L528-L535